src/dshould/prettyprint.d

```diff
@@ -0,0 +1,385 @@
+    {
+        length -= this.prefix.length;
+        length -= this.parenType.isNull ? 0 : 2;
+        if (length >= 0)
+        {
+            foreach (child; this.children)
+            {
+                length = child.lengthRemainsOf(length);
+                if (length < 0)
+                {
+                    break;
+                }
+            }
+        }
+        return length;
+    }
+}
+
+@("estimate the print length of a tree")
+unittest
+{
+    import dshould : be, less, should;
+
+    parse("Foo(Bar(Baz()), Baq())").get.lengthRemainsOf(10).should.be.less(0);
+}
+
+private enum ParenType
+{
+    paren, // ()
+    squareBracket, // []
+    curlyBracket, // {}
+}
+
+private dchar opening(const ParenType parenType)
+{
+    final switch (parenType)
+    {
+        case ParenType.paren:
+            return '(';
+        case ParenType.squareBracket:
+            return '[';
+        case ParenType.curlyBracket:
+            return '{';
+    }
+}
+
+private dchar closing(const ParenType parenType)
+{
+    final switch (parenType)
+    {
+        case ParenType.paren:
+            return ')';
+        case ParenType.squareBracket:
+            return ']';
+        case ParenType.curlyBracket:
+            return '}';
+    }
+}
+
+private QuotedText quoted(string text)
+{
+    return QuotedText(text);
+}
+
+// range over text that skips quoted strings
+private struct QuotedText
+{
+    string text; // current read head after skipping quotes
+
+    string textBeforeSkip; // current read head before skipping quotes
+
+    this(string text)
+    {
+        this(text, text);
+    }
+
+    private this(string text, string textBeforeSkip)
+    {
+        this.text = text;
+        this.textBeforeSkip = textBeforeSkip;
+        skipQuote;
+    }
+
+    // return text from start until other, which must be a different range over the same text
+    string textUntil(QuotedText other)
+    in (other.text.ptr >= this.text.ptr && other.text.ptr <= this.text.ptr + this.text.length)
+    {
+        // from our skip-front to other's skip-back
+        // ie. foo"test"bar
+        // from   ^ to ^ is the "same" range, but returns '"test"'
+        return this.textBeforeSkip[0 .. this.textBeforeSkip.length - other.text.length];
+    }
+
+    @("quote at the beginning and end of a range")
+    unittest
+    {
+        import dshould : equal, should;
+
+        auto range = QuotedText(`"Foo"`);
+
+        range.textUntil(range).should.equal(`"Foo"`);
+    }
+
+    bool empty() const
+    {
+        return this.text.empty;
+    }
+
+    dchar front() const
+    {
+        return this.text.front;
+    }
+
+    QuotedText save() const
+    {
+        return QuotedText(this.text, this.textBeforeSkip);
+    }
+
+    void popFront()
+    {
+        this.text.popFront;
+        this.textBeforeSkip = this.text;
+        skipQuote;
+    }
+
+    private void skipQuote()
+    {
+        if (!this.text.empty && this.text.front == '"')
```

Comment:
If the parse tree fails to be constructed, we just fall back to the ordinary behavior. In order to get a *wrong* parse tree, a user would have to be having two array items, using unquoted strings, each of which contains an unmatched bracket, both of which complement each other.

It is my opinion that this user is evil and deserves what they get.

Reply:
We've already seen issues with "a)" and as there is no support for Wysiwig strings, "(" in back ticks is not unlikely.

The target for `dshould`s pretty printing is `boilerplate`s `toString`. That's bad.

Reply:
Adding support for `` strings would be trivial.

The target for `dshould`s prettyprinting is any `toString` that can be automatically prettyprinted. "JSON-style expressions" are not a Funkwerk invention.

Reply:
Addendum: I feel if we support "`toString` with prettyprinting", then the large majority of extant `toString` functions will "just work", and if they don't it will be straightforward to see why. If we support "`toPrettyString`", a word that we made up, then approximately none of them will ever work. For better or worse, `toString` is the API that exists.

Reply:
PPS: would it help to farm this out into a `prettyprint` library so people can make dedicated PRs for it?

Reply:
I want a switch (`version`) for `toString` code generation, whether pretty printing is wanted or not. My favorite place for this would be `boilerplate`.

First, I use this switch for unit testing: next to nice `diff`s I get readable log messages in the case where I always look at the whole log file (without `grep`).

As soon as there is a way to identify a multi-line log message in production, I can always use this switch. But I've not yet an idea how to enable log messages for pretty printed string representations: I would need "...\n%s" with optional "\n"...

Reply:
I think the "proper" thing to do this would be to improve the toString interface so that it can take an augmented output range that can take indentation hints, ie. `openIndent`/`closeIndent`. This could then be integrated into `dshould` or `log` depending on the output device. But if we're going that way, I'd like to make a forum post about it to gather community feedback and try to get tool buy-in.